x-pack/plugins/maps/public/routing/routes/maps_app/maps_app.js

```diff
@@ -0,0 +1,489 @@
+      newFilters = filterManager.getFilters();
+      newState.filters = newFilters;
+    }
+    if (!_.isEqual(globalState.refreshInterval, refreshInterval)) {
+      newState.refreshInterval = globalState.refreshInterval;
+    }
+    if (!_.isEqual(globalState.time, time)) {
+      newState.time = globalState.time;
+    }
+    if (!_.isEmpty(newState)) {
+      this.setState(newState, () => {
+        const { query, filters, refreshInterval, time } = this.state;
+        this._appStateManager.setQueryAndFilters({
+          query: this.state.query,
+          filters: filterManager.getAppFilters(),
+        });
+        this.props.dispatchSetQuery(refreshInterval, filters, query, time);
+      });
+    }
+  }
+
+  componentWillUnmount() {
+    if (this._storeSyncUnsubscribe) {
+      this._storeSyncUnsubscribe();
+    }
+    if (this._globalSyncUnsubscribe) {
+      this._globalSyncUnsubscribe();
+    }
+    if (this._appSyncUnsubscribe) {
+      this._appSyncUnsubscribe();
+    }
+    if (this._visibleSubscription) {
+      this._visibleSubscription.unsubscribe();
+    }
+
+    // Clean up app state filters
+    const { filterManager } = getData().query;
+    filterManager.filters.forEach((filter) => {
+      if (filter.$state.store === esFilters.FilterStateStore.APP_STATE) {
+        filterManager.removeFilter(filter);
+      }
+    });
+  }
+
+  _getInitialLayersFromUrlParam() {
+    const locationSplit = window.location.href.split('?');
+    if (locationSplit.length <= 1) {
+      return [];
+    }
+    const mapAppParams = new URLSearchParams(locationSplit[1]);
+    if (!mapAppParams.has('initialLayers')) {
+      return [];
+    }
+
+    try {
+      let mapInitLayers = mapAppParams.get('initialLayers');
+      if (mapInitLayers[mapInitLayers.length - 1] === '#') {
+        mapInitLayers = mapInitLayers.substr(0, mapInitLayers.length - 1);
+      }
+      return rison.decode_array(mapInitLayers);
+    } catch (e) {
+      getToasts().addWarning({
+        title: i18n.translate('xpack.maps.initialLayers.unableToParseTitle', {
+          defaultMessage: `Initial layers not added to map`,
+        }),
+        text: i18n.translate('xpack.maps.initialLayers.unableToParseMessage', {
+          defaultMessage: `Unable to parse contents of 'initialLayers' parameter. Error: {errorMsg}`,
+          values: { errorMsg: e.message },
+        }),
+      });
+      return [];
+    }
+  }
+
+  async _updateIndexPatterns(nextIndexPatternIds) {
+    const indexPatterns = [];
+    const getIndexPatternPromises = nextIndexPatternIds.map(async (indexPatternId) => {
+      try {
+        const indexPattern = await getIndexPatternService().get(indexPatternId);
+        indexPatterns.push(indexPattern);
+      } catch (err) {
+        // unable to fetch index pattern
+      }
+    });
+
+    await Promise.all(getIndexPatternPromises);
+    this.setState({
+      indexPatterns,
+    });
+  }
+
+  _handleStoreChanges = async () => {
```

Comment:
This method is not needed. Redux already subscribes you to store changes. remove method and `this._storeSyncUnsubscribe = subscribeToSyncStore(this._handleStoreChanges);`. Then just move the below segment into componentDidUpdate

```
if (nextIndexPatternIds !== prevIndexPatternIds) {
      storeUpdates.prevIndexPatternIds = nextIndexPatternIds;
      await this._updateIndexPatterns(nextIndexPatternIds);
    }
```